Tensile/Source/client/source/HardwareMonitor.cpp

```diff
@@ -386,9 +390,178 @@ namespace Tensile
                 else
                     m_fanValues[i] += newValue;
             }
+            
+            rsmi_gpu_metrics_t gpuMetrics;
+            auto status = rsmi_dev_gpu_metrics_info_get(m_smiDeviceIndex, &gpuMetrics);
+            if(status != RSMI_STATUS_SUCCESS)
+            {
+                m_freqValues[0]        = std::numeric_limits<uint16_t>::max();
+                m_powerValues[0]       = std::numeric_limits<uint16_t>::max();
+                m_temperatureValues[0] = std::numeric_limits<uint16_t>::max();
+            }
+            else
+            {
+                m_freqValues.push_back(gpuMetrics.average_gfxclk_frequency);
+                m_powerValues.push_back(gpuMetrics.average_socket_power);
+                m_temperatureValues.push_back(gpuMetrics.temperature_hotspot);
+            }
 
             m_dataPoints++;
         }
+        
+        double HardwareMonitor::getMedianGfxFrequency()
+        {
+            assertNotActive();
+            if(m_dataPoints == 0)
+                throw std::runtime_error("No data points collected!");
+                
+            if (m_freqValues.size() != 0)
+            {
+                if(m_freqValues[0] == std::numeric_limits<uint16_t>::max())
+                    return std::numeric_limits<double>::quiet_NaN();
+
+                 size_t FrequencyMidValueIndex = m_freqValues.size() / 2;
+                 std::nth_element(m_freqValues.begin(), m_freqValues.begin()+FrequencyMidValueIndex, m_freqValues.end());
```

Comment:
Is m_freqValues sorted?

Reply:
The nth_element is in place and does not return anything.

Reply:
std:nth_element does partial sorting. it does find a mid position such that all values before this is less than the midpoint value and all values after this will be greater than this mid point value. Efficiency through this way seem to be slightly better than full sorting to find median.  Full sorting is n log(n) Vs O(n) . so I think this could be better. but can revisit later if needed.